components/table/table-wrapper.js

```diff
@@ -0,0 +1,318 @@
+				--d2l-table-header-background-color: #ffffff;
+			}
+		`;
+	}
+
+	constructor() {
+		super();
+		this.stickyHeaders = false;
+		this.type = 'default';
+		this._tableObserver = null;
+	}
+
+	disconnectedCallback() {
+		super.disconnectedCallback();
+		if (this._tableObserver) this._tableObserver.disconnect();
+	}
+
+	render() {
+		const slot = html`<slot @slotchange="${this._handleSlotChange}"></slot>`;
+		if (this.stickyHeaders) {
+			return slot;
+		} else {
+			return html`<d2l-scroll-wrapper>${slot}</d2l-scroll-wrapper>`;
+		}
+	}
+
+	updated(changedProperties) {
+		// hack: grades/groups/outcomes in the LE use this CSS class on the
+		// body to apply special CSS to the page when tables are sticky
+		// Ideally they should be adding this class to the body.
+		if (changedProperties.has('stickyHeaders')) {
+			if (this.stickyHeaders) {
+				document.body.classList.add('d2l-table-sticky-headers');
+			}
+		}
+	}
+
+	async _applyClassNames(table) {
+
+		await new Promise((resolve) => requestAnimationFrame(() => resolve()));
+
+		// offsetParent causes reflow/paint so do them all at once
+		const rows = Array.from(table.rows);
+		let firstRow = null;
+		let lastRow = null;
+		rows.forEach((r) => {
+			if (r.offsetParent !== null) {
+				if (firstRow === null) {
+					firstRow = r;
+				}
+				lastRow = r;
+			}
+		});
+
+		const topHeader = table.querySelector('tr[header]:first-child th:not([rowspan]), thead tr:first-child th:not([rowspan])');
+		const topHeaderHeight = topHeader ? topHeader.clientHeight : -1;
+
+		let prevRow = null;
+		let skipFirst = 0;
+		rows.forEach((r) => {
+
+			const isHeader = r.parentNode.tagName === 'THEAD' || r.hasAttribute('header');
+
+			let firstNonHeaderRow = !isHeader;
+			if (prevRow) {
+				const isPrevRowHeader = prevRow.parentNode.tagName === 'THEAD' || prevRow.hasAttribute('header');
+				firstNonHeaderRow = firstNonHeaderRow && isPrevRowHeader;
+				prevRow.classList.toggle('d2l-table-selected-previous', r.hasAttribute('selected') && !isPrevRowHeader);
+			}
+
+			r.classList.toggle('d2l-table-row-first', r === firstRow);
+			r.classList.toggle('d2l-table-row-last', r === lastRow);
+			r.classList.toggle('d2l-table-selected-first', firstNonHeaderRow && r.hasAttribute('selected'));
+
+			Array.from(r.cells).forEach((c, index) => {
+				c.classList.toggle('d2l-table-cell-first', index === 0 && skipFirst === 0);
+				if (index === 0 && skipFirst === 0 && c.hasAttribute('rowspan')) {
+					skipFirst = parseInt(c.getAttribute('rowspan'));
+				}
+				c.classList.toggle('d2l-table-cell-last', index === r.cells.length - 1);
+			});
+
+			prevRow = r;
+			skipFirst = Math.max(0, --skipFirst);
+
+		});
+
+		if (this.stickyHeaders && topHeaderHeight > -1) {
+			const offset = this.type === 'default' ? -3 : 1; // default: -5px top + 2px border, light: 0 top + 1px border
+			const ths = Array.from(table.querySelectorAll('tr[header]:not(:first-child) th, thead tr:not(:first-child) th'));
```

Comment:
Do we ever need to worry about nested tables?

Reply:
We've tried to worry about them in the sense that our CSS just targets the rows/cells that are directly in the `.d2l-table` table.

But you're right that this particular sticky-header query is going to find nested table rows as well. It's the same as it was (so I haven't made it worse!) at least. This approach is also pretty broken... like it works for a sticky header with two rows (like grades), but if you had 2+ rows it wouldn't work.